[SPARK-23889][SQL] DataSourceV2: required sorting and clustering for writes

aokolnychyi opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?

This PR implements the core part of the [design doc](https://docs.google.com/document/d/1X0NsQSryvNmXBY9kcvfINeYyKC-AahZarUqg3nS1GQs/edit#) for SPARK-23889.

**Note**: This PR contains all changes in one place to simplify the review. Once we agree on the approach, I am going to split it into smaller PRs.

<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Data sources should be able to request a specific distribution and ordering of data on write. In particular, these scenarios are considered useful:
- global sort
- cluster data and sort within partitions
- local sort within partitions
- no sort

Please see the design doc above for a more detailed explanation of requirements.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

This PR introduces public changes to the DS V2 by adding a logical write abstraction as we have on the read path as well as additional interfaces to represent distribution and ordering of data (please see the doc for more info).

Important pieces:
- `Write` - a logical representation of a data source write
- `RequiresDistributionAndOrdering` - a write that requires a specific distribution/ordering
- `V2Writes` - a rule that constructs a logical write and inserts repartition/sort nodes
- `WriteDistributionAndOrderingSuite` - a test case with samples

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

The patch comes with a new test case.

aokolnychyi committed:
[SPARK-23889][SQL] DataSourceV2: required sorting and clustering for writes

Lead-authored-by: Anton Okolnychyi <aokolnychyi@apple.com>
Co-authored-by: Ryan Blue <blue@apache.org>

SparkQA commented:
**[Test build #125632 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125632/testReport)** for PR 29066 at commit [`e5a42ab`](https://github.com/apache/spark/commit/e5a42ab89a314edee2feb069752feee45bd3bb01).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30245/
Test PASSed.

SparkQA commented:
**[Test build #125632 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125632/testReport)** for PR 29066 at commit [`e5a42ab`](https://github.com/apache/spark/commit/e5a42ab89a314edee2feb069752feee45bd3bb01).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class Distributions `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125632/
Test FAILed.

SparkQA commented:
**[Test build #125634 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125634/testReport)** for PR 29066 at commit [`6fea82b`](https://github.com/apache/spark/commit/6fea82b34621843844f9fde8a70b113f170c74b2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30248/
Test PASSed.

SparkQA commented:
**[Test build #125634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125634/testReport)** for PR 29066 at commit [`6fea82b`](https://github.com/apache/spark/commit/6fea82b34621843844f9fde8a70b113f170c74b2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class Distributions `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125634/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #125690 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125690/testReport)** for PR 29066 at commit [`6fea82b`](https://github.com/apache/spark/commit/6fea82b34621843844f9fde8a70b113f170c74b2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30303/
Test PASSed.

SparkQA commented:
**[Test build #125690 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125690/testReport)** for PR 29066 at commit [`6fea82b`](https://github.com/apache/spark/commit/6fea82b34621843844f9fde8a70b113f170c74b2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class Distributions `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125690/
Test FAILed.

dongjoon-hyun commented:
Hi, @aokolnychyi .
Could you fix the following UT failure?
-  org.apache.spark.sql.execution.arrow.ArrowConvertersSuite.test Arrow Validator

cc @rdblue since this is DSv2.

aokolnychyi commented:
The test failure in `ArrowConvertersSuite` is related to the optimization I did to dedup sorts. I've created [another](https://github.com/apache/spark/pull/29089) PR to address that separately. Will rebase this one afterwards.

SparkQA commented:
**[Test build #125856 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125856/testReport)** for PR 29066 at commit [`d75f0e4`](https://github.com/apache/spark/commit/d75f0e4185490a16f42131aee14618cc3653e5af).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30466/
Test PASSed.

aokolnychyi commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:
```diff
@@ -0,0 +1,102 @@
+/**
+ * A rule that constructs [[Write]]s.
+ */
+object V2Writes extends Rule[LogicalPlan] with PredicateHelper {
```
This rule contains the same logic we had before except it is applied earlier now.

aokolnychyi replied:
I think `buildAndRun` methods in exec nodes still contain the old logic. Previously, it was called `run`.

SparkQA commented:
**[Test build #125857 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125857/testReport)** for PR 29066 at commit [`5e4d304`](https://github.com/apache/spark/commit/5e4d304abbbbeed08d447bf0c8e7878810f18456).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30467/
Test PASSed.

SparkQA commented:
**[Test build #125856 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125856/testReport)** for PR 29066 at commit [`d75f0e4`](https://github.com/apache/spark/commit/d75f0e4185490a16f42131aee14618cc3653e5af).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125856/
Test FAILed.

SparkQA commented:
**[Test build #125857 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125857/testReport)** for PR 29066 at commit [`5e4d304`](https://github.com/apache/spark/commit/5e4d304abbbbeed08d447bf0c8e7878810f18456).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125857/
Test PASSed.

SparkQA commented:
**[Test build #126190 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126190/testReport)** for PR 29066 at commit [`98fb788`](https://github.com/apache/spark/commit/98fb788a9146171797dbe59f921c1eb6628720ac).

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #126190 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126190/testReport)** for PR 29066 at commit [`98fb788`](https://github.com/apache/spark/commit/98fb788a9146171797dbe59f921c1eb6628720ac).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class Distributions `
  * `case class WriteToDataSourceV2FallbackExec(`
  * `case class V2BatchWriteCommand(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126190/
Test FAILed.

aokolnychyi commented:
retest this please

SparkQA commented:
**[Test build #126195 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126195/testReport)** for PR 29066 at commit [`98fb788`](https://github.com/apache/spark/commit/98fb788a9146171797dbe59f921c1eb6628720ac).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30808/
Test PASSed.

SparkQA commented:
**[Test build #126195 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126195/testReport)** for PR 29066 at commit [`98fb788`](https://github.com/apache/spark/commit/98fb788a9146171797dbe59f921c1eb6628720ac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class Distributions `
  * `case class WriteToDataSourceV2FallbackExec(`
  * `case class V2BatchWriteCommand(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126195/
Test PASSed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #127158 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127158/testReport)** for PR 29066 at commit [`98fb788`](https://github.com/apache/spark/commit/98fb788a9146171797dbe59f921c1eb6628720ac).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31780/
Test PASSed.

SparkQA commented:
**[Test build #127158 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127158/testReport)** for PR 29066 at commit [`98fb788`](https://github.com/apache/spark/commit/98fb788a9146171797dbe59f921c1eb6628720ac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class Distributions `
  * `case class WriteToDataSourceV2FallbackExec(`
  * `case class V2BatchWriteCommand(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127158/
Test PASSed.

dongjoon-hyun commented:
Hi, @aokolnychyi .
I know that you've been waiting for a long time in WIP status.
How do you want to proceed this proposal? Is there any updates from your side?

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:
```diff
@@ -0,0 +1,102 @@
+      val write = writeBuilder.build()
+      V2BatchWriteCommand(write, query)
+
+    case OverwriteByExpression(relation: DataSourceV2Relation, deleteExpr, query, options, _) =>
+      // fail if any filter cannot be converted. correctness depends on removing all matching data.
+      val filters = splitConjunctivePredicates(deleteExpr).map {
+        filter => DataSourceStrategy.translateFilter(deleteExpr,
+          supportNestedPredicatePushdown = true).getOrElse(
+          throw new AnalysisException(s"Cannot translate expression to source filter: $filter"))
+      }.toArray
```
By the change, we move catalyst expression -> sources.Filter conversion to logical plans. So we will see both catalyst expressions and sources.expressions in logical plans in optimization.

Is it more clear if we use only catalyst expressions in logical plans, and convert to sources.Filter in physical plans when we need to interact datasources?

viirya replied:
And later in V2WriteRequirements, we also need to convert sources.Filter back to catalyst expressions.

rdblue replied:
I don't think that it is possible to do the conversion later, because we need the write builder to be fully configured to produce a `Write` in this rule. That way, the `Write` can expose its order and distribution requirements in the other rule.

I think it is fine to convert to `Filter` here. That's the public API for filter expressions, so I don't think there is a requirement for it to be used only in physical plans. Filter is already used in the optimizer on the read path as well, because the `Scan` is similarly built and added to an optimizer plan so that the optimizer can handle stats based on the pushed filters.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:
```diff
@@ -0,0 +1,102 @@
+          throw new SparkException(s"Table does not support overwrite by expression: $table")
+      }
+
+      V2BatchWriteCommand(write, query)
```
Another question is, by this change we use one single `V2BatchWriteCommand` for all kind of writing commands. But the direction after deprecated `WriteToDataSourceV2`, seems to use more specific logical plans like `AppendData`.

rdblue replied:
I can see why it made sense to move back to a generic `WriteToDataSourceV2`, but I think we do want write-specific exec nodes, even if the information isn't used.

This change would cause all v2 writes to show up in the Spark UI as `WriteToDataSourceV2`, rather than the current `AppendData` or `OverwritePartitionsDynamic`. It also makes it so the physical plan doesn't show important information, like the delete expression for `OverwriteByExpression`.

That information _could_ be recovered from the source's `Write` using getters, or maybe it would be part of the write's `toString` output, but I wouldn't want to rely on either of those. The Spark plan should keep track of this information for good `EXPLAIN` output.

aokolnychyi replied:
I am resolving this thread as the original plans are kept.

HeartSaVioR commented:
This is definitely a major missing piece on DSv2 compared to DSv1, as DSv1 writer can deal with Dataframe directly, hence able to do arbitrary changes (including repartition/sort) before doing actual write, like I did for state data source -  https://github.com/HeartSaVioR/spark-state-tools/blob/8a74bdb1bc7911a6f71785cf68b784b0a331a1d9/src/main/scala/net/heartsavior/spark/sql/state/StateStoreWriter.scala#L67-L74

Lots of data sources are blocked to migrate to DSv2. Shall we consider prioritizing this?

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/WriteBuilder.java:
```diff
@@ -23,17 +23,34 @@
+  /**
+   * Returns a logical {@link Write} shared between batch and streaming.
+   */
+  default Write build() {
```
The `buildForBatch` method (and stream equivalent) are already released, so this generic `Write` implementation makes the new structure, `build` + `toBatch`, work for existing sources. It also allows sources to implement the version that they choose. So if none of the features that require the `Write` are used, I guess they could avoid a mostly-boilerplate class.

aokolnychyi replied:
Correct, this method was introduced to keep the compatibility.

sunchao replied:
Have you considered to change the default impl for `buildForBatch` to:
```scala
  default BatchWrite buildForBatch() {
    build().toBatch()
  }
```

and also the `build()` to just return a simple anonymous `new Write() {}`? 

Otherwise, I can see that we'll have the `buildForBatch` (and similarly `buildForStreaming`) logic in two different places: `WriteBuilder` and `Write`. It is easy to miss one or another.

aokolnychyi replied:
I am not sure I understood. Could you elaborate a bit more, @sunchao?

Spark will now always call `build()` and work with the `Write` abstraction. I added the default implementation so that  existing data sources that already implement the current API will continue to work as before. Spark will is not supposed to call `buildForBatch` after this change.

aokolnychyi replied:
New data sources should be encouraged to implement only `build`.

aokolnychyi replied:
We should probably deprecate the other ones.

sunchao replied:
What I mean is now we can potentially have two copies of the `toBatch` implementation: one in `WriteBuilder.buildForBatch` and one in `Write.toBatch`, when users start to override `build`, `buildForBatch` and `buildForStreaming`. If moving forward we want `build` to be the canonical impl, perhaps we can make `buildForBatch` and `buildForStreaming` to just call `build.toBatch()` internally so that users just need to override `build`.

aokolnychyi commented:
I'll go through the comments later this week and update the PR.

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/NullOrdering.java:
```diff
@@ -0,0 +1,22 @@
+package org.apache.spark.sql.connector.expressions;
+
+public enum NullOrdering {
+  NULLS_FIRST, NULLS_LAST
```
Since this is used in `describe()` output, it would be nice to have a `toString` or `describe` method that returns the SQL-like string.

aokolnychyi replied:
I implemented `toString`.

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/SortDirection.java:
```diff
@@ -0,0 +1,22 @@
+package org.apache.spark.sql.connector.expressions;
+
+public enum SortDirection {
+  ASCENDING, DESCENDING
```
Same here. It would be nice to implement `describe`.

aokolnychyi replied:
I implemented `toString`.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:
```diff
@@ -0,0 +1,102 @@
+    case OverwriteByExpression(relation: DataSourceV2Relation, deleteExpr, query, options, _) =>
+      // fail if any filter cannot be converted. correctness depends on removing all matching data.
+      val filters = splitConjunctivePredicates(deleteExpr).map {
+        filter => DataSourceStrategy.translateFilter(deleteExpr,
```
I think the intent was to call `translateFilter` on `filter` instead of `deleteExpr`?

aokolnychyi replied:
Fixed.

holdenk commented:
I took look at the design doc, I think the design is solid. I'm happy to take a more detailed look once you are ready for code review.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2WriteRequirements.scala:
```diff
@@ -0,0 +1,107 @@
+        // this part is controversial as we perform resolution in the optimizer
+        // we cannot perform this step in the analyzer since we need to optimize expressions
+        // in nodes like OverwriteByExpression before constructing a logical write
+        query.resolve(ref.parts, resolver) match {
```
The `LogicalPlan.resolve` methods are used in several places that aren't strictly in the analyzer. I think it is fine to do this here.

A couple of examples:

* [`InsertIntoHiveTable`](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala#L194-L197) (exec node)
* [`DataSourceStrategy`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala#L210-L211)

aokolnychyi replied:
I'll be open to alternatives but I don't see a better way and we already have examples that do this.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2WriteRequirements.scala:
```diff
@@ -0,0 +1,107 @@
+          case Some(attr) => attr
+          case None => throw new AnalysisException(s"Cannot resolve '$ref' using ${query.output}")
+        }
+      case _ =>
```
I think it would be helpful to support literals and transforms that are built-in Spark functions as well. For example, a timestamp column might use `date_trunc('HOUR', ts)` to co-locate data for a given hour. This doesn't need to happen in the initial commit, but transforms will be needed fairly early on.

aokolnychyi replied:
I agree.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2WriteRequirements.scala:
```diff
@@ -0,0 +1,107 @@
+        // repartition by expression will correctly pick range or hash partitioning
+        // based on whether we have an ordered or clustered distribution
+        // all expressions in OrderedDistribution will be mapped into SortOrder
+        // all expressions in ClusteredDistribution will be mapped into Expression
```
This comment wasn't clear until I took a look at `RepartitionByExpression`. I think it may be more clear to say "the conversion to catalyst expressions above produces SortOrder for OrderedDistribution and Expression for ClusteredDistribution, and RepartitionByExpression will detect SortOrder expressions and use range partitioning."

aokolnychyi replied:
Fixed.

rdblue commented on sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala:
```diff
@@ -0,0 +1,594 @@
+    .add("data", StringType)
+
+  writeOperations.foreach { operation =>
+    test(s"ordered distribution and sort with same exprs ($operation)") {
```
Using `s" ... $var "` format strings breaks IntelliJ. I think it is better to call a test method like this:

```scala
test("case: 1") {
  testMethod(1)
}

test("case: 2") {
  testMethod(2)
}

def testMethod(param: int): Unit = { ... }
```

rdblue commented on sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala:
```diff
@@ -0,0 +1,594 @@
+  }
+}
+
+class ExtendedInMemoryTable(
```
Why not add the distribution and order to the existing `InMemoryTable` and set default args that disable sorting and distribution?

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -186,6 +186,13 @@ abstract class Optimizer(catalogManager: CatalogManager)
+    // Expression optimizations must be run before this batch so that we have optimal
+    // expressions when we construct writes. At the same time, rules that dedup repartition and
+    // sort operators must by run afterwards.
+    Batch("Early Writes", Once, earlyWriteRules: _*) :+
```
You can avoid the need to make the rules `Idempotent` by using `FixedPoint(1)` instead of `Once`. There are a few existing rules that do it, like the "Join Reorder" just below.

aokolnychyi replied:
Fixed.

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/distributions/ClusteredDistribution.java:
```diff
@@ -0,0 +1,28 @@
+ * A distribution where tuples that share the same values for clustering expressions are co-located
+ * in the same partition.
+ */
+public interface ClusteredDistribution extends Distribution {
```
I think the new interfaces should be marked `Experimental` using Spark's annotations, too.

aokolnychyi replied:
I added the annotations but yet to add docs.

rdblue commented:
I'm testing this out and I hit a problem. There is no way to create Spark's internal SortOrder implementation that is public. I created a class that implements SortOrder, but that fails in `toCatalyst`: `Caused by: java.lang.RuntimeException: com.netflix.spark.OrderField@7e1d244 is not currently supported`.

Although the interface is defined, this doesn't accept any implementation of that interface. For the public `Expression` API, we added extractors that can match any implementation of the public interfaces. I think we need to do that for the `SortOrder` as well. And, we should expose a way to create a `SortOrder` expression. Here's an example extractor: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala#L192-L199

rdblue commented:
Here's an extractor that should work:

SparkQA commented:
**[Test build #131719 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131719/testReport)** for PR 29066 at commit [`d47817d`](https://github.com/apache/spark/commit/d47817ddb185041d49af84131a53c33877ec00e4).

aokolnychyi commented:
Sorry, it took a while to update this PR.

@rdblue had time to play around with this change locally and addressed quite some review comments. I rebased, incorporated the changes and added @rdblue as a co-author.

I am going to fix the remaining minor points next but the PR is ready for review.

aokolnychyi commented:
Okay, I went through the comments and I think they are all resolved except points related to tests. This PR is no longer WIP and is ready for a detailed review.

aokolnychyi commented:
I'd like to emphasize that all changes are in one place to simplify the review. I'll split the work into smaller PRs later.

SparkQA commented:
**[Test build #131726 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131726/testReport)** for PR 29066 at commit [`ecc6a46`](https://github.com/apache/spark/commit/ecc6a463f3d6c8147e84d7067b7e3d457167f475).

SparkQA commented:
**[Test build #131726 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131726/testReport)** for PR 29066 at commit [`ecc6a46`](https://github.com/apache/spark/commit/ecc6a463f3d6c8147e84d7067b7e3d457167f475).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131726/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36317/

SparkQA commented:
**[Test build #131719 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131719/testReport)** for PR 29066 at commit [`d47817d`](https://github.com/apache/spark/commit/d47817ddb185041d49af84131a53c33877ec00e4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class Distributions `

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131719/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36324/

aokolnychyi commented:
Tests failed as I overlooked recent changes around caching. Should be fixed now.

SparkQA commented:
**[Test build #131791 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131791/testReport)** for PR 29066 at commit [`f7e8444`](https://github.com/apache/spark/commit/f7e84440b2eeabb383ab34b63f3bc93f428654b0).

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36389/

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -189,6 +189,13 @@ abstract class Optimizer(catalogManager: CatalogManager)
+    // Expression optimizations must be run before this batch so that we have optimal
+    // expressions when we construct writes. At the same time, rules that dedup repartition and
+    // sort operators must by run afterwards.
+    Batch("Early Writes", Once, earlyWriteRules: _*) :+
```
Should we combine `earlyScanPushDownRules` and `earlyWriteRules`? I think they are compatible. We could just come up with a better name for this phase, like `v2SourceRewriteRules`.

aokolnychyi replied:
I think they may be more rules like this in the future, not just writes. I am definitely +1 on making this more flexible. Let me do this in a separate PR.

aokolnychyi replied:
I am resolving this as this was implemented in #30558.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:
```diff
@@ -0,0 +1,187 @@
+    case overwrite @ OverwriteByExpression(
+        relation: DataSourceV2Relation, deleteExpr, query, options, _, None) =>
+      // fail if any filter cannot be converted. correctness depends on removing all matching data.
+      val filters = splitConjunctivePredicates(deleteExpr).flatMap { p =>
```
Nit: `p` isn't a very good variable name. What about `pred`?

aokolnychyi replied:
I think I named it to stay on one line, I'll double check.

rdblue commented:
This looks ready to me. We've been running with these changes in prod and they work great.

SparkQA commented:
**[Test build #131791 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131791/testReport)** for PR 29066 at commit [`f7e8444`](https://github.com/apache/spark/commit/f7e84440b2eeabb383ab34b63f3bc93f428654b0).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait V2ExistingTableWriteExec extends V2TableWriteExec `
  * `trait V2TableWriteExec extends V2CommandExec with UnaryExecNode `

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131791/

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #131837 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131837/testReport)** for PR 29066 at commit [`f7e8444`](https://github.com/apache/spark/commit/f7e84440b2eeabb383ab34b63f3bc93f428654b0).

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36434/

SparkQA commented:
**[Test build #131837 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131837/testReport)** for PR 29066 at commit [`f7e8444`](https://github.com/apache/spark/commit/f7e84440b2eeabb383ab34b63f3bc93f428654b0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait V2ExistingTableWriteExec extends V2TableWriteExec `
  * `trait V2TableWriteExec extends V2CommandExec with UnaryExecNode `

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131837/

HeartSaVioR commented:
Great to see this making progress. It'd be nice if we include this in Spark 3.1.0, so that custom V2 writers could leverage this instead of sticking with V1 writer.
(Personally I think it worths to ensure including this in 3.1.0. This shouldn't drag any more, discussion happened years ago, and V2 writer also exists years ago and has been lacking such essential functionality.)

I tried to take a look when the PR was WIP, but I felt I'm not quite qualified to review and finally approve this. I'll try taking a look soon, but it'd be nice if some others can review this.

cc. @cloud-fan @viirya Would you mind taking a look at this to make this in 3.1.0? Thanks in advance.

aokolnychyi commented:
I also have a prototype for this logic in micro-batch streaming. I added dedicated plans which I think we were missing for a while. Right now, `MicroBatchExecution` replaces `WriteToMicroBatchDataSource` with `WriteToDataSourceV2`, a deprecated node, in `runBatch`. I don't think we use that path anywhere except tests but it would be great to get that done.

aokolnychyi commented:
also cc @dbtsai @dongjoon-hyun, it would be great to get your input on this one after the holidays.

aokolnychyi commented:
We should agree on the future of the existing `Distribution` and `ClusteredDistribution` interfaces used in `Partitioning`.

Here is a quote from the design doc:

> Currently, Distribution (defined as part of public DS V2) is specific to the read path and is not flexible (e.g. it does not allow transforms). Therefore, the proposal is to make it more generic and reuse in the write path.

I am interested in what other devs think and whether we are OK breaking the existing API. It seems inevitable to me at some point (e.g. no way to use partition transforms in the existing interfaces).

HeartSaVioR commented:
From what I understand is replacing existing `Distribution` and `ClusteredDistribution` in read path (package `org.apache.spark.sql.connector.read.partitioning`) with new addition here, otherwise the new addition has to be only used for write path, and read and write paths are ended up diverged. Do I understand correctly?

I'm not sure how many data sources in ecosystem are already taking down steps to follow up DSv2 changes in Spark 3.0.0. These existing interfaces are marked as `Evolving` so they're allowed to make change, though I'm not 100% sure that is also OK to unify, which means both package and signature are changing. Probably worth to raise a discussion in dev@ mailing list?

Btw, the information on distribution and sort order is for optimization (avoid shuffle) on read path, but on write path it's not just an optimization but could be "requirements" for specific data sources, blocking specific data sources to implement. I personally feel this should have higher priority than the read path, unfortunately.

rdblue commented:
> I am interested in what other devs think and whether we are OK breaking the existing API.

Since the other API is targeted at the read path, I would have no problem adding this one in parallel under a `write` package. I think that we should deprecate the read-side distribution because it doesn't really help with bucketed joins.

I'm also fine changing the existing API, but I'd rather just deprecate it and remove it when we have a replacement for bucketed joins and other read-side optimizations.

> Probably worth to raise a discussion in dev@ mailing list?

Yes. But if we want to get this into 3.1.0, we should start moving on everything in parallel. We should start getting the addition of `Write` done because it needs to carry the `RequiresDistributionAndSort` interface no matter what we decide about `Distribution`. And we can at least get a WIP PR up to add the new distribution interfaces.

dbtsai commented:
@aokolnychyi can you rebase master to resolve the conflict? 

@sunchao @cloud-fan @gatorsmile could you help to take a look? 

Thanks.

sunchao commented:
Yes I'm looking this and will have review ready today or tomorrow. Thanks.

HeartSaVioR commented:
I'm OK to go in parallel - as I said write path is more critical for the functionality so read path shouldn't block this.

The main point here may be "where" we should put the new interfaces. If I understand correctly, the direction of the PR (and the question) is rather we unify the interfaces between read path and write path, assuming the new interfaces could cover the further read path optimization as well. (I'm rather novice here, hence no opinion on the coverage.)

It also sounds a good approach if we go with separate package for separate path, and let them evolve separately, and eventually merge once they've stabilized. The downside would be that it would require more changes on downstream during evolution.

aokolnychyi commented:
@dbtsai, I will rebase this one once PR #30558 is in.

aokolnychyi commented:
I know deprecating and then removing is usually a better idea and I will be okay evolving read and write path separately. The only concern I have is that while we use these interfaces in the write path here, the concept isn't really write-specific.

sunchao commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/distributions/ClusteredDistribution.java:
```diff
@@ -0,0 +1,30 @@
+ */
+@Experimental
+public interface ClusteredDistribution extends Distribution {
+  Expression[] clustering();
```
NVM if you think this is obvious :)

aokolnychyi replied:
Oh, yeah, docs and annotations should be added for sure. I'll fix while rebasing.

aokolnychyi replied:
I added more docs. There may be places where we would want a bit more details but I think we can review those parts when I split this PR into smaller chunks.

sunchao commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:
```diff
@@ -0,0 +1,187 @@
+      val write = writeBuilder match {
+        case builder: SupportsTruncate if isTruncate(filters) =>
+          builder.truncate().build()
+        case builder: SupportsOverwrite =>
```
Should we put `SupportsOverwrite` before `SupportsTruncate`? if a builder class extends `SupportsOverwrite` (and `isTruncate` returns true) then it will be matched by the first clause and call the `truncate` method, but we want it to call `overwrite(filters)` right?

rdblue replied:
This is existing logic, just moved.

The reason for this is that `SupportsOverwrite` extends `SupportsTruncate` and calls `overwrite(true)`. Calling `truncate` ensures that the source can implement either one if it chooses. Sometimes truncate may be preferred, and it is easier for a source to receive that call directly rather than writing its own equivalent of `isTruncate`.

sunchao replied:
Got it. I'm just not sure if some data source would extend `SupportsOverwrite` and decides to override/implement different behavior for `overwrite` and `truncate` (which no longer calls `overwrite(true)`, and here we'd pick `truncate` but what they really want is `overwrite`.

rdblue replied:
If the source doesn't implement `truncate`, then it gets `overwrite(true)`: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/SupportsOverwrite.java#L46-L48

aokolnychyi replied:
This logic did not change and should match the previous behavior.

sunchao reviewed (commented):
Sorry for the sporadic review comments. This PR looks mostly good to me and the main thing concerning me is the newly introduced `Distribution` interface and how that evolve with the existing one in `read` package. Happy to see discussions already happened on this and agree that we can move start moving in parallel.

aokolnychyi commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/SortOrder.java:
```diff
@@ -0,0 +1,27 @@
+
+import org.apache.spark.annotation.Experimental;
+
+@Experimental
```
I am not sure whether this will be part of 3.1.0. Once we have clarity, I'll add the annotation.

aokolnychyi commented on sql/catalyst/src/main/scala/org/apache/spark/sql/connector/distributions/distributions.scala:
```diff
@@ -0,0 +1,59 @@
+}
+
+private[sql] final case class ClusteredDistributionImpl(
+    clusteringExprs: Seq[Expression]) extends ClusteredDistribution {
```
I switched to using `Seq` in fields to avoid reasoning about equality of arrays.

aokolnychyi commented on sql/core/src/main/java/org/apache/spark/sql/connector/write/V1Write.java:
```diff
@@ -0,0 +1,26 @@
+import org.apache.spark.sql.sources.InsertableRelation;
+
+@Unstable
+public interface V1Write extends Write {
```
Added a bit of description.

aokolnychyi commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V1FallbackWriters.scala:
```diff
@@ -38,10 +38,11 @@ case class AppendDataExecV1(
+    override val write: Option[V1Write] = None) extends V1FallbackWriters {
 
-  override protected def run(): Seq[InternalRow] = {
-    writeWithV1(newWriteBuilder().buildForV1Write(), refreshCache = refreshCache)
```
I moved `refreshCache` to parent.

aokolnychyi commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:
```diff
@@ -0,0 +1,187 @@
+ * This rule does resolution in the optimizer because some nodes like [[OverwriteByExpression]]
+ * must undergo the expression optimization before we can construct a logical write.
+ */
+object V2Writes extends Rule[LogicalPlan] with PredicateHelper {
```
This rule not only inserts shuffle/sort but also build `Write`s. It is only applied if `Write` has not been constructed before.

aokolnychyi commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -216,12 +216,12 @@ case class AppendDataExec(
 
-  override protected def run(): Seq[InternalRow] = {
-    val writtenRows = writeWithV2(newWriteBuilder().buildForBatch())
-    refreshCache()
```
Refresh happens in `V2ExistingTableWriteExec` now.

aokolnychyi commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -322,6 +322,27 @@ trait BatchWriteHelper {
+trait V2ExistingTableWriteExec extends V2TableWriteExec {
+  def session: SparkSession
+  def relation: DataSourceV2Relation
+  def write: Option[BatchWrite] = None
```
I kept it as `Option` just in case we don't want to apply the new logic all the time and will introduce a flag to fallback to the old approach. If we are not going to have that flag, we can make this required.

sunchao replied:
Cool, it would simplify the logic if we know the new code will always be applied. Perhaps worth creating a JIRA to track this.

aokolnychyi commented:
I've updated this PR and I am ready to split it into smaller mergeable parts. It would be great if everyone could take another look to make sure we are on the same page.

aokolnychyi commented:
Seems like there is consensus about evolving this API alongside the interfaces in `read` package. I am not sure whether we need to move new interfaces to `write`, though. This concept isn't really read or write specific.

SparkQA commented:
**[Test build #132373 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132373/testReport)** for PR 29066 at commit [`32f5687`](https://github.com/apache/spark/commit/32f5687fd628a4ffca38fa0840fd52447db7d680).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class Distributions `
  * `trait V2ExistingTableWriteExec extends V2TableWriteExec `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36974/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/132373/

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:
```diff
@@ -0,0 +1,185 @@
+        // the conversion to catalyst expressions above produces SortOrder expressions
+        // for OrderedDistribution and generic expressions for ClusteredDistribution
+        // this allows RepartitionByExpression to pick either range or hash partitioning
+        RepartitionByExpression(distribution, query, numShufflePartitions)
```
Is it possible the required distribution be changed later by other optimization? The distribution requirement from data source is a hard requirement? Once if the distribution is changed and not matched the requirement, how will data source react to it?

aokolnychyi replied:
We are inserting repartition/sort nodes directly before writing so my assumption that Spark will only remove them if the incoming plan already satisfies these requirements. `WriteDistributionAndOrderingSuite` is kind of meant for testing that. Do you have ideas when this assumption will break, @viirya?

viirya replied:
Oh, I see. Mis-reading this part. Looks fine. Thanks.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:
```diff
@@ -0,0 +1,185 @@
+      resolver: Resolver): catalyst.expressions.Expression = {
+    def resolve(ref: FieldReference): NamedExpression = {
+      // this part is controversial as we perform resolution in the optimizer
+      // we cannot perform this step in the analyzer since we need to optimize expressions
+      // in nodes like OverwriteByExpression before constructing a logical write
```
If we resolve it in the analyzer, cannot we optimize the resolved expressions later in the optimizer?

aokolnychyi replied:
At this step, we construct a `Write` and pass the overwrite expressions to the data source. Expression optimization must have happened before.

viirya replied:
Got it. Thanks for clarifying.

aokolnychyi commented on sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:
```diff
@@ -261,3 +340,20 @@ private[sql] object FieldReference {
   }
 }
+
+private[sql] final case class SortValue(
```
I am open to alternatives here for sure. It seems like we are giving synonyms in this file. For example, `FieldReference` implements `NamedReference`. Unfortunately, I cannot use `SortOrder` as it is already taken in the public expression API.

aokolnychyi commented on sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala:
```diff
@@ -0,0 +1,594 @@
+
+  // TODO: do we need to dedup repartitions too? RepartitionByExpr -> Projects -> RepartitionByExpr
+  writeOperations.foreach { operation =>
+    ignore(s"ordered distribution and sort with manual repartition ($operation)") {
```
I am not sure whether it is safe to do. @dongjoon-hyun @viirya, what's your take on this?

aokolnychyi commented:
It is a bit hard to keep this large PR up-to-date since it touches many places. As it seems like a reasonable approach, I am going to split the work and submit smaller PRs. We can perform detailed reviews on individual PRs.

aokolnychyi commented:
The first PR with interfaces only is out.

HyukjinKwon commented:
@aokolnychyi, so https://github.com/apache/spark/pull/30706 and https://github.com/apache/spark/pull/30577 were the splits? Yeah, I think splitting is a good approach for a big change like this. Should we close this PR BTW?

aokolnychyi commented:
Closing this one in favor of smaller PRs.
